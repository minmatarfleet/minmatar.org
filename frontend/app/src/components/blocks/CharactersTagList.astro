---
import { i18n } from '@helpers/i18n'
const { t, translatePath } = i18n(Astro.url)

import type { CharacterTag } from '@dtypes/api.minmatar.org'
import type { CharacterTagSummary } from '@dtypes/layout_components'

interface Props {
    tags:               CharacterTag[];
    tags_summary:       CharacterTagSummary[];
    active_character:   number;
}

const {
    tags,
    tags_summary,
    active_character,
} = Astro.props

const active_tags = tags_summary.find(character => character.character_id === active_character)?.tags

const REEL_GAP = 20

import { get_player_icon } from '@helpers/eve_image_server'

import Flexblock from '@components/compositions/Flexblock.astro';
import Wrapper from '@components/compositions/Wrapper.astro';
import Grid from '@components/compositions/Grid.astro';

import StylessButton from '@components/blocks/StylessButton.astro';
import SelectableCard from '@components/blocks/SelectableCard.astro'
import CharactersReel from '@components/blocks/CharactersReel.astro';
import CharacterReelSlide from '@components/blocks/CharacterReelSlide.astro';
import Badge from '@components/blocks/Badge.astro';
---

<Flexblock id="character-tag-list" gap='var(--space-2xl)'>
    <CharactersReel gap={REEL_GAP} initial_slide={tags_summary.findIndex(character => character.character_id === active_character)}>
        {tags_summary.map(character =>
            <CharacterReelSlide gap={REEL_GAP}>
                <Wrapper padding_block='0' padding_inline='0' max_width='350px'>
                    {active_character === character.character_id ?
                        <Badge
                            character_id={character.character_id}
                            title={character.character_name}
                            image={get_player_icon(character.character_id)}
                            title_el="h2"
                        >
                            <small>{`${character.tags.length} ${character.tags.length === 1 ? t('role_selected') : t('roles_selected')}`}</small>
                        </Badge>
                        :
                        <StylessButton
                            class="w-full character-selector"
                            href={translatePath(`/account/tags/${character.character_id}`)}
                            x-data={`{
                                redirect(accepted) {
                                    if (accepted) navigate("${translatePath(`/account/tags/${character.character_id}`)}")
                                },
                                show_unsaved_dialog() {
                                    show_confirm_dialog({
                                        title: '${t('unsaved_changes_dialog_title')}',
                                        content: '${t('unsaved_changes_dialog_text')}',
                                    }).then( (accepted) => this.redirect(accepted) )
                                }
                            }`}
                            x-on:click.prevent="is_dirty ? show_unsaved_dialog() : redirect(true)"
                        >
                            <Badge
                                character_id={character.character_id}
                                title={character.character_name}
                                image={get_player_icon(character.character_id)}
                                title_el="h2"
                            >
                                <small>{`${character.tags.length} ${character.tags.length === 1 ? t('role_selected') : t('roles_selected')}`}</small>
                            </Badge>
                        </StylessButton>
                    }
                        
                </Wrapper>
            </CharacterReelSlide>
        )}
    </CharactersReel>

    <Grid
        role="list"
        class="[ w-full grid-fill ]"
        min_item_width='240px'
        row_gap='var(--space-3xs)'
        column_gap='var(--space-3xs)'
    >
        {tags.map(tag =>
            <SelectableCard
                image={tag.image_name}
                label={tag.description}
                name="tag"
                value={tag.id}
                checked={active_tags?.includes(tag.id)}
                x-on:change="is_dirty = true; console.log(is_dirty)"
            >
                <Flexblock gap='var(--space-3xs)'>
                    <h4>{t(`${tag.description}_tag` as any)}</h4>
                    <small>{t(`${tag.description}_tag_description` as any)}</small>
                </Flexblock>
            </SelectableCard>    
        )}
    </Grid>
</Flexblock>

<style lang="scss">
    .character-selector {
        opacity: 0.5;

        &:hover,
        &:focus-visible {
            opacity: 0.8;
        }

        &.active {
            opacity: 1;
        }

        h2 {
            color: var(--faded);
        }

        small {
            color: var(--fleet-yellow);
        }
    }
</style>