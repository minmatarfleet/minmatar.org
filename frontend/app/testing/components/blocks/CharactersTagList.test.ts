import { experimental_AstroContainer as AstroContainer } from "astro/container";
import { expect, test, vi } from "vitest";
import CharactersTagList from "@components/blocks/CharactersTagList.astro";

test("CharactersTagList defaults", async () => {
  const container = await AstroContainer.create();
  const result = await container.renderToString(CharactersTagList, {
    props: {
      tags: [
        {
          id: 4523,
          image_name: "some-image",
          description: "the sky is falling",
        },
      ],
      tags_summary: [
        {
          character_id: 18,
          tags: [],
        },
        {
          character_id: 2345,
          tags: [],
        },
      ],
      active_character: {
        character_id: 2345,
      },
    },
  });

  expect(result).toMatchSnapshot();
});
