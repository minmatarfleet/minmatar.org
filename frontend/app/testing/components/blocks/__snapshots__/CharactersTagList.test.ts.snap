// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharactersTagList defaults 1`] = `
"<div class="flex-block gap---space-2xl" id="character-tag-list" data-astro-cid-5xcyectv="true" data-mod style="--flex_block_gap: var(--space-2xl);">  <div class="wrapper [ staging-reel ]" data-astro-cid-vmvp644o="true" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: 0;--wrapper_padding_inline_end: 0;--wrapper_padding_block: 0;--wrapper_max_width: 100%;">  <div class="[ swiper ][ w-full ]" data-astro-cid-vmvp644o> <div class="flex-inline [ swiper-wrapper ][ select-none ]" x-data="{
        swiper: null,
        corporation_pagination(index) {
            return this.pagination_icons[index]
        },
        init() {
            this.swiper = new Swiper('.swiper', {
                initialSlide: -1,
                slidesPerView: &#34;auto&#34;,
                spaceBetween: 20,
                pagination: false,
                loop: false,
                fadeEffect: {
                    crossFade: true
                },
                navigation: {
                    nextEl: '.swiper-button-next',
                    prevEl: '.swiper-button-prev',
                },
            });
        }
    }" data-astro-cid-okd446ho="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: 0;--flex_inline_justification: flex-start;">   <div class="[ swiper-slide ]" style="margin-right: 20px" data-astro-cid-pvvgnag4>  <div class="wrapper" data-astro-cid-5xcyectv="true" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: 0;--wrapper_padding_inline_end: 0;--wrapper_padding_block: 0;--wrapper_max_width: 350px;"> <a href="/account/tags/18" class="w-full character-selector" x-data="{
                                redirect(accepted) {
                                    if (accepted) navigate(&#34;/account/tags/18&#34;)
                                },
                                show_unsaved_dialog() {
                                    show_confirm_dialog({
                                        title: 'Stay on Page',
                                        content: 'You have unsaved changes on this page. If you leave now, your character roles will not be updated. Are you sure you want to continue without saving?',
                                    }).then( (accepted) => this.redirect(accepted) )
                                }
                            }" x-on:click.prevent="is_dirty ? show_unsaved_dialog() : redirect(true)" data-astro-cid-5xcyectv="true" data-astro-cid-syldxee3="true">  <div class="fixed-fluid [ items-center ]" character_id="18" data-astro-cid-5xcyectv="true" data-astro-cid-etghgjam="true" data-mod data-astro-cid-34go4nvk style="--fixed_fluid_gap: var(--space-s);--fixed_fluid_width: 64px;--fixed_fluid_breakpoint: 50%;">   <picture data-astro-cid-etghgjam> <img loading="lazy" src="https://images.evetech.net/characters/18/portrait?size=64" width="64" height="64" data-astro-cid-etghgjam> </picture>  <div class="flex-block gap-0" data-astro-cid-etghgjam="true" data-mod style="--flex_block_gap: 0;">  <h2 data-astro-cid-etghgjam="true" class="min-w-[200px] text-[var(--highlight)]"> </h2>  <small data-astro-cid-5xcyectv>0 roles selected</small>   </div>   </div>     </a>  </div>    </div> <div class="[ swiper-slide ]" style="margin-right: 20px" data-astro-cid-pvvgnag4>  <div class="wrapper" data-astro-cid-5xcyectv="true" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: 0;--wrapper_padding_inline_end: 0;--wrapper_padding_block: 0;--wrapper_max_width: 350px;"> <a href="/account/tags/2345" class="w-full character-selector" x-data="{
                                redirect(accepted) {
                                    if (accepted) navigate(&#34;/account/tags/2345&#34;)
                                },
                                show_unsaved_dialog() {
                                    show_confirm_dialog({
                                        title: 'Stay on Page',
                                        content: 'You have unsaved changes on this page. If you leave now, your character roles will not be updated. Are you sure you want to continue without saving?',
                                    }).then( (accepted) => this.redirect(accepted) )
                                }
                            }" x-on:click.prevent="is_dirty ? show_unsaved_dialog() : redirect(true)" data-astro-cid-5xcyectv="true" data-astro-cid-syldxee3="true">  <div class="fixed-fluid [ items-center ]" character_id="2345" data-astro-cid-5xcyectv="true" data-astro-cid-etghgjam="true" data-mod data-astro-cid-34go4nvk style="--fixed_fluid_gap: var(--space-s);--fixed_fluid_width: 64px;--fixed_fluid_breakpoint: 50%;">   <picture data-astro-cid-etghgjam> <img loading="lazy" src="https://images.evetech.net/characters/2345/portrait?size=64" width="64" height="64" data-astro-cid-etghgjam> </picture>  <div class="flex-block gap-0" data-astro-cid-etghgjam="true" data-mod style="--flex_block_gap: 0;">  <h2 data-astro-cid-etghgjam="true" class="min-w-[200px] text-[var(--highlight)]"> </h2>  <small data-astro-cid-5xcyectv>0 roles selected</small>   </div>   </div>     </a>  </div>    </div>    </div>   <div class="swiper-pagination" data-astro-cid-okd446ho></div>  </div> <div class="[ swiper-button-prev hidden lg:block ]" data-astro-cid-vmvp644o></div> <div class="[ swiper-button-next hidden lg:block ]" data-astro-cid-vmvp644o></div>  </div>    <div class="grid [ w-full grid-fill ]" role="list" data-astro-cid-5xcyectv="true" data-astro-cid-fkxmsq5x style="--grid_row_gap: var(--space-3xs);--grid_column_gap: var(--space-3xs);--grid_min_item_width: 240px;"> <label class="[ cursor-pointer ]" data-astro-cid-umgfmho4> <input type="checkbox" name="tag" value="4523" x-on:change="is_dirty = true; console.log(is_dirty)" data-astro-cid-5xcyectv="true" class="[ selectable-card ] astro-umgfmho4" data-astro-cid-umgfmho4>  <div class="wrapper [ card component ] [ card card-animated ] light-transparency" data-astro-cid-idiuzidr="true" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: 0;--wrapper_padding_inline_end: 0;--wrapper_padding_block: 0;--wrapper_max_width: 100%;">   <div class="flex-block gap-0" data-astro-cid-idiuzidr="true" data-mod style="--flex_block_gap: 0;">  <picture class="animated" data-astro-cid-bcfgy4kl> <img class="mask" loading="lazy" height="412" width="256" src="some-image" data-astro-cid-bcfgy4kl> </picture>  <div class="wrapper [ card-body ]" data-astro-cid-idiuzidr="true" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: var(--space-2xs);--wrapper_padding_inline_end: var(--space-2xs);--wrapper_padding_block: var(--space-2xs);--wrapper_max_width: 80rem;">        <div class="flex-block gap---space-3xs" data-astro-cid-5xcyectv="true" data-mod style="--flex_block_gap: var(--space-3xs);">  <h4 data-astro-cid-5xcyectv></h4> <small data-astro-cid-5xcyectv></small>  </div>      </div>   <div class="wrapper [ card-header ]" data-astro-cid-idiuzidr="true" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: var(--space-2xs);--wrapper_padding_inline_end: var(--space-2xs);--wrapper_padding_block: var(--space-2xs);--wrapper_max_width: 80rem;">    </div>   <div class="wrapper [ card-actions ]" data-astro-cid-idiuzidr="true" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: var(--space-2xs);--wrapper_padding_inline_end: var(--space-2xs);--wrapper_padding_block: var(--space-2xs);--wrapper_max_width: 80rem;">    </div>    </div>    </div>     </label>   </div>    </div>  "
`;
